node/src/components/consensus/highway_core/validators.rs

```diff
@@ -209,16 +228,22 @@ impl<T> ValidatorMap<T> {
 
     /// Binary searches this sorted `ValidatorMap` for `x`.
     ///
-    /// If the value is found, `Ok` is returned, containing the index, otherwise `Err`, with the
-    /// first index at which the value is greater than `x`.
-    pub fn binary_search(&self, x: &T) -> Result<ValidatorIndex, ValidatorIndex>
+    /// Returns the lowest index of an entry `>= x`, or `None` if `x` is greater than all entries.
+    pub fn binary_search(&self, x: &T) -> Option<ValidatorIndex>
     where
         T: Ord,
     {
-        self.0
-            .binary_search(x)
-            .map(|i| ValidatorIndex(i as u32))
-            .map_err(|i| ValidatorIndex(i as u32))
+        match self.0.binary_search(x) {
+            Ok(i) => Some(ValidatorIndex(
+                (0..i)
+                    .rev()
```

Comment:
Why is the reversion necessary?

Reply:
If `x` occurs multiple times, [`slice::binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) will return an index of any occurrence, not necessarily the first. We want to select a random validator and guarantee that someone with weight 0 is excluded, so we want to go back to the _first_ occurrence of `x` in the list of cumulative weights.

E.g. if `x` is at index 100, 101 and 102 (because validators 101 and 102 have weight 0 or are excluded), and `slice::binary_search` returns 102, we want to go back to 100. This will usually be much faster than starting at 0 until we reach 100 (then we wouldn't need a binary search in the first place: we'd just do a linear search).

Reply:
Thanks for the explanation. It's definitely not clear from the implementation. I think `binary_search` term is already so strongly rooted in people's minds that overloading it like you're doing here is dangerous. As you explained above, what we're doing here makes sense if `T = Weight` and not necessarily for other `T`s. Can you move this method to be defined on `Validators<Weight>` only and maybe change the name?

Reply:
Why? It's a general binary search that returns the _first_ hit instead of an unspecified one, and that returns `None` if `x` is greater than all entries.

It just turns out that that's the kind of binary search we want for our random selection.